src/lib/storage/column_encoders/dictionary_encoder.hpp

```diff
@@ -0,0 +1,108 @@
+#pragma once
+
+#include <algorithm>
+#include <limits>
+#include <memory>
+
+#include "base_column_encoder.hpp"
+
+#include "storage/encoded_columns/dictionary_column.hpp"
+#include "storage/value_column.hpp"
+#include "storage/zero_suppression/utils.hpp"
+#include "types.hpp"
+#include "utils/enum_constant.hpp"
+
+namespace opossum {
+
+class DictionaryEncoder : public ColumnEncoder<DictionaryEncoder> {
+ public:
+  static constexpr auto _encoding_type = enum_c<EncodingType, EncodingType::Dictionary>;
+
+  template <typename T>
+  std::shared_ptr<BaseColumn> _encode(const std::shared_ptr<ValueColumn<T>>& value_column) {
+    // See: https://goo.gl/MCM5rr
+    // Create dictionary (enforce uniqueness and sorting)
+    const auto& values = value_column->values();
+    auto dictionary = pmr_vector<T>{values.cbegin(), values.cend(), values.get_allocator()};
+
+    // Remove null values from value vector
+    if (value_column->is_nullable()) {
+      const auto& null_values = value_column->null_values();
+
+      // Swap values to back if value is null
+      auto erase_from_here_it = dictionary.end();
+      auto null_it = null_values.crbegin();
+      for (auto dict_it = dictionary.rbegin(); dict_it != dictionary.rend(); ++dict_it, ++null_it) {
+        if (*null_it) {
+          std::swap(*dict_it, *(--erase_from_here_it));
+        }
+      }
+
+      // Erase null values
+      dictionary.erase(erase_from_here_it, dictionary.end());
+    }
+
+    std::sort(dictionary.begin(), dictionary.end());
+    dictionary.erase(std::unique(dictionary.begin(), dictionary.end()), dictionary.end());
+    dictionary.shrink_to_fit();
+
+    auto attribute_vector = pmr_vector<uint32_t>{values.get_allocator()};
```

Comment:
So we are first creating a standard-size vector and will compress it later? This should be documented somewhere.

Reply:
Okay

Reply:
done